### PR TITLE
Properly clear metadata and other structs when reusing ServerContext in callback API

### PR DESCRIPTION
**Pull request by vjpai**
Credit for test and issue identification to @hcaseyal

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +0.0%     +24 [None]                           +5.18Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +622%    +454 grpc::ServerContext::Setup          +454  +622%

  +0.1%    +472 TOTAL                            +5.62Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +0.0%     +24 [None]                           +5.18Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +622%    +454 grpc::ServerContext::Setup          +454  +622%

  +0.1%    +472 TOTAL                            +5.62Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,028      Total (=)      2,019,028

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,179,738      Total (>)     11,179,727

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +0.0%     +24 [None]                           +5.18Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +622%    +454 grpc::ServerContext::Setup          +454  +622%

  +0.1%    +472 TOTAL                            +5.62Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +0.0%     +24 [None]                           +5.18Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +622%    +454 grpc::ServerContext::Setup          +454  +622%

  +0.1%    +472 TOTAL                            +5.62Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,028      Total (=)      2,019,028

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,179,735      Total (<)     11,179,738

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.6%     +96 src/core/lib/surface/server.cc                                                           +96  +0.6%
      [NEW]    +146 request_matcher_zombify_all_pending_calls((anonymous namespace)::request_matcher*) [    +146  [NEW]
       +37%     +82 init_call_elem                                                                           +82   +37%
      +3.8%     +14 [Unmapped]                                                                               +14  +3.8%

 -+-+-+-+-+-+-+ MIXED                                                                                +-+-+-+-+-+-+-
  -0.0%     -48 [None]                                                                                  +104  +0.0%

  +0.0%     +48 TOTAL                                                                                   +200  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +0.0%     +24 [None]                           +5.18Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +622%    +454 grpc::ServerContext::Setup          +454  +622%

  +0.1%    +472 TOTAL                            +5.62Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,028      Total (=)      2,019,028

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,179,740      Total (>)     11,179,737

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,028      Total (=)      2,019,028

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,179,731      Total (<)     11,179,737

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

File: src/cpp/server/server_context.cc [hunk 1]
```diff
@@ -233,6 +233,10 @@ void ServerContext::Setup(gpr_timespec deadline) {
   call_ = nullptr;
   cq_ = nullptr;
   sent_initial_metadata_ = false;
+  auth_context_.reset();
```
Comment (hcaseyal) [inline src/cpp/server/server_context.cc]: Should this be moved to Clear(), rather than having it in Setup()?

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  [ = ]       0 [None]                           +7.48Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +210%    +445 grpc::ServerContext::Clear          +445  +210%
      +2.2%      +3 [Unmapped]                            +3  +2.2%

  +0.1%    +448 TOTAL                            +7.91Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  [ = ]       0 [None]                           +7.48Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +210%    +445 grpc::ServerContext::Clear          +445  +210%
      +2.2%      +3 [Unmapped]                            +3  +2.2%

  +0.1%    +448 TOTAL                            +7.91Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,380      Total (=)      2,019,380

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,180,294      Total (<)     11,180,303

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,380      Total (=)      2,019,380

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,180,308      Total (>)     11,180,298

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by vjpai** [comment]
Remaining failures unrelated

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  [ = ]       0 [None]                           +7.48Ki  +0.1%
  +3.7%    +448 src/cpp/server/server_context.cc    +448  +3.7%
      +210%    +445 grpc::ServerContext::Clear          +445  +210%
      +2.2%      +3 [Unmapped]                            +3  +2.2%

  +0.1%    +448 TOTAL                            +7.91Ki  +0.1%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,404      Total (=)      2,019,404

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,180,297      Total (<)     11,180,303

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Pull request merged**